packages/adblocker-electron/adblocker.ts

```diff
@@ -7,17 +27,136 @@ export function fromElectronDetails({
+    ses.setPreloads([join(__dirname, './content.js')]);
+
+    ipcMain.on('is-mutation-observer-enabled', (event: Electron.IpcMessageEvent) => {
+      event.returnValue = this.options.mutationObserver;
+    });
+  }
+
+  private onGetCosmeticFilters = (
+    event: Electron.IpcMessageEvent,
+    url: string,
+    msg: IBackgroundCallback & { action?: string },
+  ): void => {
+    // Extract hostname from sender's URL
+    const parsed = parse(url);
+    const hostname = parsed.hostname || '';
+    const domain = parsed.domain || '';
+
+    const { active, styles, scripts } = this.getCosmeticsFilters({
+      domain,
+      hostname,
+      url,
+
+      classes: msg.classes,
+      hrefs: msg.hrefs,
+      ids: msg.ids,
+
+      // This needs to be done only once per frame
+      getBaseRules: msg.lifecycle === 'start',
+      getInjectionRules: msg.lifecycle === 'start',
+      getRulesFromHostname: msg.lifecycle === 'start',
+
+      // This will be done every time we get information about DOM mutation
+      getRulesFromDOM: msg.lifecycle === 'dom-update',
+    });
+
+    if (active === false) {
+      return;
+    }
+
+    // Inject custom stylesheets
+    this.injectStyles(event.sender, styles);
+
+    // Inject scripts from content script
+    event.sender.send('get-cosmetic-filters-response', {
+      active,
+      extended: [],
+      scripts,
+      styles: '',
+    } as IMessageFromBackground);
+  }
+
+  private onHeadersReceived = (
+    details: Electron.OnHeadersReceivedDetails,
+    callback: any,
+  ): void => {
+    const policies: string[] = [];
+    const responseHeaders: any = details.responseHeaders || {};
+
+    if (details.resourceType === 'mainFrame' || details.resourceType === 'subFrame') {
+      const CSP_HEADER_NAME = 'content-security-policy';
+      const rawCSP = this.getCSPDirectives(fromElectronDetails(details));
+      if (rawCSP !== undefined) {
+        policies.push(...(rawCSP === undefined ? [] : rawCSP.split(';').map(csp => csp.trim())));
+
+        // Collect existing CSP headers from response
+        for (const [name, value] of Object.entries(responseHeaders)) {
+          if (name.toLowerCase() === CSP_HEADER_NAME) {
+            policies.push(value as string);
+            responseHeaders[name] = undefined;
+          }
+        }
+
+        responseHeaders['Content-Security-Policy'] = policies;
+
+        callback({ responseHeaders });
+        return;
+      }
+    }
+
+    callback({});
+  }
+
+  private onBeforeRequest = (
+    details: Electron.OnBeforeRequestDetails,
+    callback: (a: Electron.Response) => void,
+  ): void => {
+    const { redirect, match } = this.match(fromElectronDetails(details));
+
+    if (redirect) {
+      callback({ redirectURL: redirect.dataUrl });
+    } else if (match) {
+      callback({ cancel: true });
+    } else {
+      callback({});
+    }
+  }
+
+  private injectStyles(sender: Electron.WebContents, styles: string): void {
+    if (sender) {
```

Comment:
In which case can `sender` be falsy?

Reply:
Well, that's a safe check whether it's not null or undefined.